### PR TITLE
test: test case for issue #1120

### PR DESCRIPTION
**Pull request by PopGoesTheWza**
Sample test to reproduce issue #1120 

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [ ] If it's a new feature, I have included documentation updates.

### CODE REVIEW

**Comment by PopGoesTheWza** [comment]
@szmarczak @sindresorhus please comment this (very) kludgy fix.

In short, `.buffer()`, `.json()` and `.text()` shortcuts consume the `body` Buffer bound to the initial `asPromise()` call and cannot "see" the other `body` bound to the `asPromise()` which is called during the `afterResponse` hook.

There is very likely a cleaner way to handle this issue.

File: source/as-promise.ts [hunk 1]
```diff
@@ -7,17 +7,17 @@ import {normalizeArguments, mergeOptions} from './normalize-arguments';
 import requestAsEventEmitter, {proxyEvents} from './request-as-event-emitter';
 import {CancelableRequest, GeneralError, NormalizedOptions, Response} from './types';
 
-const parseBody = (body: Buffer, responseType: NormalizedOptions['responseType'], encoding: NormalizedOptions['encoding']): unknown => {
+const parseBody = (kludge: {body: Buffer}, responseType: NormalizedOptions['responseType'], encoding: NormalizedOptions['encoding']): unknown => {
```
Comment (szmarczak) [inline source/as-promise.ts]: don't do this

Comment (PopGoesTheWza) [reply source/as-promise.ts]: @szmarczak please elaborate. My intent wasn't to commit a proper fix but to hint at what and how this issue occurs.

I never expected this kludge to pollute the codebase and couldn't decide on a better approach.

Comment (PopGoesTheWza) [reply source/as-promise.ts]: Perhaps a better resolution would be to:

- call the `shortcut()` factory function with a reference to the `promise` variable
- then from within `shortcut()` to bind the resulting buffer()/json()/text() function to said `promise` reference
- finally when invoked, buffer()/json()/text() function will `await` its `this` in order to access the `response`, next the `body`

I can give it a try if you think that approach is acceptable.

**Comment by PopGoesTheWza** [comment]
@szmarczak here is a cleaner fix (type casting is somewhat ugly though)

**Comment by szmarczak** [comment]
If you don't mind waiting until I finish up the Rewrite PR, it will be much easier to look into this later...

**Comment by PopGoesTheWza** [comment]
Sure.

**Comment by sindresorhus** [comment]
This can continue now. https://github.com/sindresorhus/got/pull/1051 was merged.

**Comment by PopGoesTheWza** [comment]
Let me now if and how I may assist

**Comment by szmarczak** [comment]
Thanks for the reproducible example, fixed in 99d70df02f209ffac80a6529001e63b8684056af

**Pull request closed without merging**